### PR TITLE
*DO NOT MERGE* feat(CodeGen): commits to add codegen support for native modules.

### PR DESCRIPTION
**Pull request by rozele**

### CODE REVIEW

**Comment by rozele** [comment]
I'm not convinced about this yet - it probably won't hurt to merge as is, the only impact on perf is a type check for `INativeModuleWrapper` in a couple places and a slightly larger NPM package because of the `ReactNative.CodeGen` project (O(100KB)). 

I'd love to find a way that this could be easily integrated into the build process, the problem is that I couldn't find a way to reflect on UWP assemblies except from a UWP class library, and you can't run UWP class libraries from the command line.

File: ReactWindows/ReactNative.Shared/ReactInstanceManager.cs [hunk 1]
```diff
@@ -713,10 +713,17 @@ private void DetachViewFromInstance(ReactRootView rootView, IReactInstance react
...
+            var codeGenReactPackage = new CodeGen.MetadataExtractorReactPackage(reactPackage);
+            reactPackage = codeGenReactPackage;
+#endif
             foreach (var nativeModule in reactPackage.CreateNativeModules(reactContext))
             {
                 nativeRegistryBuilder.Add(nativeModule);
             }
+#if RNCODEGEN
+            CodeGen.MetadataWriter.Write(codeGenReactPackage.Metadata);
+#endif
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/ReactInstanceManager.cs]: Let's encapsulate this in MetadataExtractor for now.

**Comment by matthargett** [comment]
Can you document the app startup performance improvement here that you mentioned on Reactiflux? That’ll help other reviewers better evaluate the cost/benefit :) thanks!

**Comment by rozele** [comment]
@antonkh has a few bugs to report on this. This will have to be rebased off #1788

**Comment by rozele** [comment]
Closing out of date PR

**Pull request closed without merging**